### PR TITLE
Fix #6742: Only possible to build station next to competitor by using CTRL+click

### LINKED ISSUE #6742: Only possible to build station next to competitor by using CTRL+click
Hello, 

I don't know if it's actually a bug or not, but I discovered that it is possible to build a rail station directly next to other player airport (see attached screenshot) using ctrl+click (building a station with possibility of combining two stations into one).

It is **not possible** to build station directly by just clicking (without ctrl), because that way an popup pops up saying that: 
**Can't build railroad station here...
Adjoins more than one existing station/loading area**

So I think it should not be allowed even using ctrl+click. If I am wrong, just close/delete this issue. :)

adjoined_heli_and_train

Version 1.8.0, playing on Rpi 1/b+, custom build (without any changes to the source code).

### PR DESCRIPTION
**Pull request by Hemaolle**
This builds on @Hopman's original pull request (#6831) for by adding a fix suggested in that pull request for the case where there are two competitor stations nearby.

Closes #6831

### CODE REVIEW

**Comment by LordAro** [comment]
I feel like the code would be cleaner if `GetOwnedStationAt` took the company ID as a parameter, rather than _current_company being hard coded in

Also, "there are no more than" (is->are)

**Comment by LordAro** [comment]
As for commit messages, should probably be a Remove followed by a Fix, given those are the first words following the update :p

(Only need the issue number referenced once as well)

**Comment by Hemaolle** [comment]
Can you clarify the comment about the commit messages? Are you referring to a specific commit message out of the 3 here? Do you mean that the commits themselves should be in a different order? Or are you referring to the message on the pull request perhaps?

**Comment by Hemaolle** [comment]
I removed now the duplicated issue number from the PR message at least.

**Comment by LordAro** [comment]
The commit messages themselves - "Update" is rarely used, and when it's immediately followed by another keyword (Fix, Remove) you can probably do a minor amount of rewording to use that instead. The order is fine

**Comment by Hemaolle** [comment]
And only the first commit should have the issue number?

**Comment by LordAro** [comment]
It doesn't really matter - the name of the PR itself should be enough

**Comment by Hemaolle** [comment]
Addressed everything else besides "there are no more than" (is->are). For that I read: https://english.stackexchange.com/questions/35389/there-is-are-more-than-one-whats-the-difference and it seems to me based on that that my version would be correct.

**Review by LordAro — DISMISSED** [review]
Ah, the English language, always with the edge cases (I say this as a native speaker)

I wonder about using the CompanyByte type instead of OwnerByte, but since they're identical and I don't know the different uses of them (there's probably some existing (mis)use of them anyway), I won't worry about it

File: src/station_cmd.cpp [hunk 1]
```diff
@@ -89,25 +89,25 @@ bool IsHangar(TileIndex t)
...
  * @param ta the area to search over
- * @param closest_station the closest station found so far
+ * @param closest_station the closest owned station found so far
+ * @param owner the company whose stations to look for
  * @param st to 'return' the found station
  * @return Succeeded command (if zero or one station found) or failed command (for two or more stations found).
  */
 template <class T>
-CommandCost GetStationAround(TileArea ta, StationID closest_station, T **st)
+CommandCost GetStationAround(TileArea ta, StationID closest_station, OwnerByte owner, T **st)
```
Comment (michicc) [inline src/station_cmd.cpp]: Just a tiny nit-pick: The `...Byte` types are usually only used for save/load. In normal functions, `Owner` is preferred, even if just for aesthetics.

Comment (Hemaolle) [reply src/station_cmd.cpp]: Changed this to CompanyID now, perhaps that's a bit more appropriate here. What do you think?

**Comment by Hemaolle** [comment]
Changed the parameter new parameter to `GetStationAround` still from `OwnerByte` to `CompanyID`. How does that look?

**Comment by frosch123** [comment]
Are the first two commits needed? 
To me it looks like the third one fixes everything on its own.

**Comment by Hemaolle** [comment]
@frosch123, you are probably right in that the third commit would fix the issue on its own. Nevertheless, I think the 2 other commits are good cleanup since `BuildStationPart` won't anymore be passed other companies' stations to attach to. For that reason it's kind of redundant to check in that method if the passed station is owned by the current company actually.

**Comment by Hemaolle** [comment]
@LordAro would you mind reviewing this again. The previous review got discarded when I addressed michicc's comment.

**Review by nielsmh — CHANGES_REQUESTED** [review]
Seems to work correctly, however please merge the two code changes and do a force-push, so there's just one "Fix bug" and one "Remove unused string" commit.

**Review by nielsmh — DISMISSED** [review]

**Comment by Hemaolle** [comment]
@nielsmh, done.

While at it, I removed the check for the `Station **st` parameter in `BuildStationPart()` being owned by the `_current_company` and replaced with an assert in the beginning as we won't be calling it anymore with other companies' stations to attach to.

File: src/station_cmd.cpp [hunk 2]
```diff
@@ -671,14 +671,12 @@ static void UpdateStationSignCoord(BaseStation *st)
  */
 static CommandCost BuildStationPart(Station **st, DoCommandFlag flags, bool reuse, TileArea area, StationNaming name_class)
 {
+	assert(*st == NULL || (*st)->owner == _current_company);
```
Comment (frosch123) [inline src/station_cmd.cpp]: I think this assertion can be triggered by a malicious client to crash the server.
When distant-joining an existing station, the client sents the station id of the station to join.
At this point the id has been checked for existence, but not for ownership.

So, this must not be an "assert(...)" but a "if (..) return CMD_ERROR;"

**Review by frosch123 — CHANGES_REQUESTED** [review]

File: src/station_cmd.cpp (hunk 2 above)
Comment (Hemaolle) [reply src/station_cmd.cpp]: Good thinking, looks like that could be the case indeed. Fixed it.

**Review by frosch123 — APPROVED** [review]

**Pull request merged**